pilot/cmd/istioctl/log.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+
+	"github.com/spf13/cobra"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"istio.io/istio/pkg/log"
+)
+
+type component struct {
+	Name      string
+	Container string
+	Label     string
+}
+
+var (
+	controlPlaneComponents = map[string]component{
+		"pilot":   {Name: "pilot", Container: "discovery", Label: "istio=pilot"},
+		"mixer":   {Name: "mixer", Container: "mixer", Label: "istio=mixer"},
+		"ingress": {Name: "ingress", Container: "", Label: "istio=ingress"},
+		"ca":      {Name: "ca", Container: "", Label: "istio=istio-ca"},
+	}
+
+	// TODO - allow retrieval of proxy logs when using istioctl logs proxy <podID> [-n namespace]
+	logCmd = &cobra.Command{
+		Use:   "logs <component> [container]",
+		Short: "Retrieves logs from an Istio component [kube only]",
+		Long: `
+Retrieves logs from an Istio component when running in Kubernetes
+	
+Supported components are:
+
+	[pilot mixer ingress ca]
+
+		`,
+		Example: `
+	# Retrieve the logs to Pilot's primary container (discovery)
+	istioctl logs pilot
```

Comment:
This example doesn't seem like a good idea, what does primary container mean? In my current setup we scale pilot to 3 pods. how do i know which pilot logs i am getting?

Reply:
I did have a think about this. What about the following?
```
istioctl logs pilot <podID> # Handles HA pilot use case
istioctl logs pilot # Looks up pod ID for you and if it receives more than 1 prompts for user input
```